internal/lockfilescalibr/translation.go

```diff
@@ -0,0 +1,199 @@
+	"package-lock.json":           "javascript/packagelockjson",
+	"pom.xml":                     "java/pomxml",
+	"buildscript-gradle.lockfile": "java/gradlelockfile",
+	"gradle.lockfile":             "java/gradlelockfile",
+	"verification-metadata.xml":   "java/gradleverificationmetadataxml",
+	"poetry.lock":                 "python/poetrylock",
+	"Pipfile.lock":                "python/Pipfilelock",
+	"pdm.lock":                    "python/pdmlock",
+	"requirements.txt":            "python/requirements",
+	"Cargo.lock":                  "rust/Cargolock",
+	"composer.lock":               "php/composerlock",
+	"mix.lock":                    "erlang/mixlock",
+	"renv.lock":                   "r/renvlock",
+	"packages.lock.json":          "dotnet/packageslockjson",
+	// "conan.lock":                  "cpp/conanlock",
+	"go.mod":       "go/gomod",
+	"Gemfile.lock": "ruby/gemfilelock",
+}
+
+// ExtractWithExtractor attempts to extract the file at the given path with the extractor passed in
+func ExtractWithExtractor(ctx context.Context, localPath string, ext filesystem.Extractor) ([]*extractor.Inventory, error) {
+	info, err := os.Stat(localPath)
+	if err != nil {
+		return nil, err
+	}
+
+	si, err := createScanInput(localPath, info)
+	if err != nil {
+		return nil, err
+	}
+	inv, err := ext.Extract(ctx, si)
+	if err != nil {
+		return nil, fmt.Errorf("(extracting as %s) %w", ext.Name(), err)
+	}
+
+	for i := range inv {
+		inv[i].Extractor = ext
+	}
+
+	return inv, nil
+}
+
+// Extract attempts to extract the file at the given path
+//
+// Args:
+//   - localPath: the path to the lockfile
+//   - extractAs: the name of the lockfile format to extract as (Using OSV-Scanner V1 extractor names)
+//
+// Returns:
+//   - []*extractor.Inventory: the extracted lockfile data
+//   - error: any errors encountered during extraction
+//
+// If extractAs is not specified, then the function will attempt to
+// identify the lockfile format based on the file name.
+//
+// If no extractors are found, then ErrNoExtractorsFound is returned.
+func Extract(ctx context.Context, localPath string, extractAs string) ([]*extractor.Inventory, error) {
+	info, err := os.Stat(localPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if extractAs != "" {
+		for _, ext := range lockfileExtractors {
+			if lockfileExtractorMapping[extractAs] == ext.Name() {
+				si, err := createScanInput(localPath, info)
+				if err != nil {
+					return nil, err
+				}
+
+				inv, err := ext.Extract(ctx, si)
+				if err != nil {
+					return nil, fmt.Errorf("(extracting as %s) %w", extractAs, err)
+				}
+
+				for i := range inv {
+					inv[i].Extractor = ext
+				}
+
+				return inv, nil
+			}
+		}
+
+		return nil, fmt.Errorf("%w, requested %s", ErrExtractorNotFound, extractAs)
+	}
+
+	output := []*extractor.Inventory{}
+	extractorFound := false
+
+	for _, ext := range lockfileExtractors {
+		if ext.FileRequired(localPath, info) {
+			extractorFound = true
```

Comment:
a lot of this looks very similar to lines 125-139. Could we factor this out into a common function?

Reply:
Done